apps/emqx/src/emqx_shared_sub.erl

```diff
@@ -404,9 +468,12 @@ init([]) ->
     {atomic, PMon} = mria:transaction(?SHARED_SUB_SHARD, fun ?MODULE:init_monitors/0),
     ok = emqx_utils_ets:new(?SHARED_SUBSCRIBER, [protected, bag]),
     ok = emqx_utils_ets:new(?ALIVE_SHARED_SUBSCRIBERS, [protected, set, {read_concurrency, true}]),
-    ok = emqx_utils_ets:new(?SHARED_SUBS_ROUND_ROBIN_COUNTER, [
+    ok = emqx_utils_ets:new(?SHARED_SUBS_DISPATCH_COUNTER, [
         public, set, {write_concurrency, true}
     ]),
+    emqx_hooks:put(
+        'client.disconnected', {?MODULE, on_client_disconnected, [self()]}, ?HP_LOWEST
```

Comment:
sorry for not thinking it though during previous reviews.
I am afraid this is not quite reliable because a client process can be killed (or crash), in which case, it has no chance to call the hooks.

We would need to monitor the publisher clients, however it would be too expensive to keep track of the monitoring references because there could potentially millions of publisher clients vs only a few subscribers.

Also, in case of clustering, nodes may have different view of which one is least loaded, so I am unsure if this is really going to be more balanced than random or hash based strategy.

Reply:
weighted random might work.
e.g. bump a counter for each pick, for `sticky_prioritize_leastload` strategy, pick with condition `Rand > 1 - (PickCount+1) / (TotalCount+N)`.
Maybe reset the counters every 10 seconds to reflect real time load.